lib/services/post.ts

```diff
@@ -0,0 +1,155 @@
+import { Post, PostStatus, User } from "@prisma/client";
+import { prisma } from "./db";
+import { deleteImage, uploadImage } from "./image";
+
+export type PostWithUser = Post & {
+  user: User;
+};
+
+export type CreatablePost = {
+  price: number;
+  description: string;
+  image: Buffer;
+  bookId: string;
+  userId: string;
+};
+
+export type UpdatablePost = {
+  price?: number;
+  description?: string;
+  image?: Buffer;
+  status?: PostStatus;
+};
+
+type UpdatablePostResolved = {
+  price?: number;
+  description?: string;
+  imageUrl?: string;
+  status?: PostStatus;
+};
+
+/**
+ * Returns the post with the given id.
+ *
+ * @param id the id of the post
+ * @param includeUser whether or not the user associated with the post should be returned in the data
+ * @returns the post with the given id
+ * @throws if the post with the given id cannot be found
+ */
+export async function getPost(
+  id: string,
+  includeUser: boolean
+): Promise<Post | PostWithUser> {
+  return prisma.post.findUnique({
+    where: {
+      id: id,
+    },
+    include: {
+      user: includeUser,
+    },
+    rejectOnNotFound: true,
+  });
+}
+
+/**
+ * Creates a post from the given information, then returns the post.
+ *
+ * Uploads the given image.
+ *
+ * @param creatablePost the information used to create the posts
+ * @returns the created post
+ */
+export async function createPost(creatablePost: CreatablePost): Promise<Post> {
+  const imageUrl = await uploadImage(creatablePost.image);
+  try {
+    return prisma.post.create({
+      data: {
+        bookId: creatablePost.bookId,
+        description: creatablePost.description,
+        imageUrl: imageUrl,
+        price: creatablePost.price,
+        userId: creatablePost.userId,
+      },
+    });
+  } catch (e) {
+    deleteImage(imageUrl);
+    throw e;
+  }
+}
+
+/**
+ * Updates the given post with the new information, then returns the post.
+ *
+ * If a new image is provided, then old image is deleted, and the new image is uploaded.
+ *
+ * @param id the id of the post to update
+ * @param updatablePost the new information for the post
+ * @returns the updated post
+ * @throws if there is no post with the given id
+ */
+export async function updatePost(
+  id: string,
+  updatablePost: UpdatablePost
+): Promise<Post> {
+  if (Boolean(updatablePost.image)) {
+    const currentPost: Post = (await getPost(id, false)) as Post;
+    const currentImageUrl = currentPost.imageUrl;
+    const newImageUrl = await uploadImage(updatablePost.image as Buffer);
+    const updatablePostResolved = getUpdatablePostResolved(
+      updatablePost,
+      newImageUrl
+    );
+
+    try {
+      const newPost = await prisma.post.update({
+        where: {
+          id: id,
+        },
+        data: updatablePostResolved,
+      });
+      if (Boolean(currentImageUrl)) {
+        await deleteImage(currentImageUrl as string);
```

Comment:
Unsafe type cast (does the if statement not make this unnecessary for the TypeScript compiler?)